sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -189,8 +189,15 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
 
   override def createDatabase(
       dbDefinition: CatalogDatabase,
-      ignoreIfExists: Boolean): Unit = withClient {
-    client.createDatabase(dbDefinition, ignoreIfExists)
+      ignoreIfExists: Boolean): Unit = {
+    try {
+      withClient {
+        client.createDatabase(dbDefinition, ignoreIfExists)
+      }
+    } catch {
+      case e: AnalysisException if e.message.contains("already exists") =>
+        throw new DatabaseAlreadyExistsException(dbDefinition.name)
```

Comment:
@cloud-fan I intercepted the exception here. looks a bit hacky?

Reply:
shall we put this logic in `withClient`?

Reply:
Two concerns if we move the logic to `withClient`:
1. How can we guarantee `org.apache.hadoop.hive.metastore.api.AlreadyExistsException` is thrown by `createDatabase` but not for other calls like `createTable`?
2. Since db name is not available, we need to parse out the db name from the message.

Are you OK with these?

Reply:
You made a good point. But it's still a bit hacky that we wrap the exception multiple times. How about
```
def withClient(convertException: Throwable => AnalysisException, body: => T) ...
def withClient(body: => T) = withClient(e => throw new AnalysisException..., body)
```

Then we can call the new overload of `withClient` in `createDatabase` and translate the hive exception directly. We can probably apply it in more places like `createTable` when we need to unify the error message further.

Reply:
Good idea! Let me try this approach.

Reply:
Updated as suggested